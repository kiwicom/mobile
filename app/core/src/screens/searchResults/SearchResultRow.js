// @flow

import * as React from 'react';
import { graphql, createFragmentContainer } from 'react-relay';
import { Text, View } from 'react-native';
import { Date, SimpleCard } from '@kiwicom/react-native-app-common';

import AirlineLogo from '../../components/flights/AirlineLogo';
import Duration from '../../components/flights/Duration';
import Price from '../../components/flights/Price';
import RouteStop from '../../components/flights/RouteStop';

import type { SearchResultRow_node } from './__generated__/SearchResultRow_node.graphql';
import type { Navigation } from '../../types/Navigation';

type Props = {|
  navigation: Navigation,
  node: SearchResultRow_node,
|};

export const SearchResultRowWithoutData = ({ node, navigation }: Props) => {
  const { duration, price, legs, departure, arrival } = node;
  const { navigate } = navigation;
  function goToWebBooking() {
    navigate('WebBooking', { bookingUrl: node.bookingUrl });
  }

  if (legs) {
    return (
      <SimpleCard onPress={goToWebBooking}>
        <View style={{ flex: 16, flexDirection: 'row' }}>
          <View style={{ flex: 2 }}>
            <Price data={price} />
          </View>

          <View style={{ flex: 14, flexDirection: 'column' }}>
            <View style={{ flexDirection: 'row' }}>
              <View style={{ flexDirection: 'column' }}>
                <RouteStop data={departure} />
                <Date dateTime={departure && departure.localTime} />
              </View>

              <Text> &rarr; </Text>

              <View style={{ flexDirection: 'column' }}>
                <RouteStop data={arrival} />
                <Date dateTime={arrival && arrival.localTime} />
              </View>
            </View>

            <View style={{ flexDirection: 'row' }}>
              {legs.map(leg => {
                if (leg) {
                  return <AirlineLogo data={leg.airline} key={leg.id} />;
                }
              })}
              <Text>
                in <Duration minutes={duration} />
              </Text>
            </View>
          </View>
        </View>
      </SimpleCard>
    );
  } else {
    // TODO: log such an event?
    return <Text>Flight legs could not be loaded (missing data).</Text>;
  }
};

export default createFragmentContainer(
  SearchResultRowWithoutData,
  graphql`
    fragment SearchResultRow_node on Flight {
      duration
      price {
        ...Price
      }
      departure {
        localTime
        ...RouteStop
      }
      arrival {
        localTime
        ...RouteStop
      }
      legs {
        id
        airline {
          ...AirlineLogo
        }
      }
      bookingUrl
    }
  `,
);
