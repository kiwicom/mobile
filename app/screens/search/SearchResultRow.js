// @flow

import * as React from 'react';
import { graphql, createFragmentContainer } from 'react-relay';
import { Text, View } from 'react-native';

import AirlineLogo from '../../components/flights/AirlineLogo';
import Date from '../../components/visual/datetime/Date';
import Duration from '../../components/flights/Duration';
import Price from '../../components/flights/Price';
import RouteStop from '../../components/flights/RouteStop';
import SimpleCard from '../../components/visual/cards/SimpleCard';

import type { SearchResultRow_node } from './__generated__/SearchResultRow_node.graphql';

type Props = {
  node: SearchResultRow_node,
};

export const SearchResultRowWithoutData = ({ node }: Props) => {
  const { duration, price, legs, departure, arrival } = node;
  if (legs) {
    return (
      <SimpleCard>
        <View style={{ flex: 16, flexDirection: 'row' }}>
          <View style={{ flex: 2 }}>
            <Price data={price} />
          </View>

          <View style={{ flex: 14, flexDirection: 'column' }}>
            <View style={{ flexDirection: 'row' }}>
              <View style={{ flexDirection: 'column' }}>
                <RouteStop data={departure} />
                <Date dateTime={departure && departure.localTime} />
              </View>

              <Text> &rarr; </Text>

              <View style={{ flexDirection: 'column' }}>
                <RouteStop data={arrival} />
                <Date dateTime={arrival && arrival.localTime} />
              </View>
            </View>

            <View style={{ flexDirection: 'row' }}>
              {legs.map(leg => {
                if (leg) {
                  return <AirlineLogo data={leg.airline} key={leg.id} />;
                }
              })}
              <Text>
                in <Duration minutes={duration} />
              </Text>
            </View>
          </View>
        </View>
      </SimpleCard>
    );
  } else {
    // TODO: log such an event?
    return <Text>Flight legs could not be loaded (missing data).</Text>;
  }
};

export default createFragmentContainer(
  SearchResultRowWithoutData,
  graphql`
    fragment SearchResultRow_node on Flight {
      duration
      price {
        ...Price
      }
      departure {
        localTime
        ...RouteStop
      }
      arrival {
        localTime
        ...RouteStop
      }
      legs {
        id
        airline {
          ...AirlineLogo
        }
      }
    }
  `,
);
