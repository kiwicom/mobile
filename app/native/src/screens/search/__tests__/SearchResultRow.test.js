// @flow

import * as React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';

import { SearchResultRowWithoutData } from '../SearchResultRow';

const renderer = new ShallowRenderer();

const navigation = {
  navigate: (screen: string, parameters?: Object) => {}, // eslint-disable-line no-unused-vars
  state: { params: {} },
};

it('renders without legs', () => {
  renderer.render(
    <SearchResultRowWithoutData
      node={{
        duration: null,
        price: null,
        departure: null,
        arrival: null,
        legs: null,
        bookingUrl: null,
      }}
      navigation={navigation}
    />,
  );
  expect(renderer.getRenderOutput()).toMatchSnapshot();
});

it('renders with partial legs', () => {
  renderer.render(
    <SearchResultRowWithoutData
      node={{
        duration: 60, // minutes
        price: {},
        departure: {
          localTime: '2017-11-19T09:00:00',
        },
        arrival: {
          localTime: '2017-11-19T12:00:00',
        },
        legs: [
          {
            id: '1',
            airline: {},
          },
          null, // simulates API failure
        ],
        bookingUrl:
          'https://www.kiwi.com/en/booking?passengers=1-0-0&price=27&token=TjFsU9KOyjpEDqm5dNSrO1ZjuRMVMgZrERrSppD325T6P8%2BP9w468E%2B3gYirNKm0xjFEaQ8brFkJXYHcFnHHQCLzHIOgXdSALYw%2B1E9gBHy4iROEATMv%2BVOrhnavl9TJOQ6nUMKttD9Kp%2BnYVHS%2Bg%2BH3eDwpD2tmBv6kMs4%2FvZ2T%2BaiEPwDOanzNDV9ssOBFTS5LrsxAgTuj91xEFd0RVNXVmGBeOjHfa9FMBnTq%2FMKRaTNBd0qsaAQ2H8Zv7zYi%2FqhMUowidxu%2BJOXCGEIm0ER6bCkglC4mea830Cu%2F9r4p%2BWUmQP1cLzJL65sPae20ZzBxroK27yQ6sH9djwQu7lq%2BmuBvhOTxNSRwE9VNJhEr6%2B3%2FjBnOiP7RRKMBlSxIF7EXsDzqY6qRtySPDRCFvSlBTtSuK8zOAO0g%2FQzq%2ByOoH0scl0yUJnVcsKyx0HsOnUQHaW4CoqZZxZrAItSVZNOvB8VuloUzWIa4Tq5wM9mBFwFbyfYIAm27%2Ba2rznXqHXyQSdefryqbhFqzBqqArueqlyZU0Y40EJe%2Fiz96yxg%3D',
      }}
      navigation={navigation}
    />,
  );
  expect(renderer.getRenderOutput()).toMatchSnapshot();
});
