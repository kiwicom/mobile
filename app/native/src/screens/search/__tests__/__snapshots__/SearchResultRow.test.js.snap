// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders with partial legs 1`] = `
<SimpleCard>
  <View
    style={
      Object {
        "flex": 16,
        "flexDirection": "row",
      }
    }
  >
    <View
      style={
        Object {
          "flex": 2,
        }
      }
    >
      <Relay(PriceWithoutData)
        data={Object {}}
      />
    </View>
    <View
      style={
        Object {
          "flex": 14,
          "flexDirection": "column",
        }
      }
    >
      <View
        style={
          Object {
            "flexDirection": "row",
          }
        }
      >
        <View
          style={
            Object {
              "flexDirection": "column",
            }
          }
        >
          <Relay(RouteStopWithoutData)
            data={
              Object {
                "localTime": "2017-11-19T09:00:00",
              }
            }
          />
          <DateComponent
            dateTime="2017-11-19T09:00:00"
          />
        </View>
        <Text
          accessible={true}
          allowFontScaling={true}
          disabled={false}
          ellipsizeMode="tail"
        >
           â†’ 
        </Text>
        <View
          style={
            Object {
              "flexDirection": "column",
            }
          }
        >
          <Relay(RouteStopWithoutData)
            data={
              Object {
                "localTime": "2017-11-19T12:00:00",
              }
            }
          />
          <DateComponent
            dateTime="2017-11-19T12:00:00"
          />
        </View>
      </View>
      <View
        style={
          Object {
            "flexDirection": "row",
          }
        }
      >
        <Relay(AirlineLogoData)
          data={Object {}}
        />
        <Text
          accessible={true}
          allowFontScaling={true}
          disabled={false}
          ellipsizeMode="tail"
        >
          in 
          <Duration
            minutes={60}
          />
        </Text>
      </View>
    </View>
  </View>
</SimpleCard>
`;

exports[`renders without legs 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  disabled={false}
  ellipsizeMode="tail"
>
  Flight legs could not be loaded (missing data).
</Text>
`;
